core/src/main/scala/kafka/server/AbstractAsyncFetcher.scala

```diff
@@ -0,0 +1,669 @@
+      }
+    }
+
+    updateFetchOffsetAndMaybeMarkTruncationComplete(fetchOffsets)
+  }
+
+  private def maybeTruncateToEpochEndOffsets(fetchedEpochs: Map[TopicPartition, EpochEndOffset],
+                                             latestEpochsForPartitions: Map[TopicPartition, EpochData]): ResultWithPartitions[Map[TopicPartition, OffsetTruncationState]] = {
+    val fetchOffsets = mutable.HashMap.empty[TopicPartition, OffsetTruncationState]
+    val partitionsWithError = mutable.HashSet.empty[TopicPartition]
+
+    fetchedEpochs.foreach { case (tp, leaderEpochOffset) =>
+      leaderEpochOffset.error match {
+        case Errors.NONE =>
+          val offsetTruncationState = getOffsetTruncationState(tp, leaderEpochOffset)
+          if(doTruncate(tp, offsetTruncationState))
+            fetchOffsets.put(tp, offsetTruncationState)
+
+        case Errors.FENCED_LEADER_EPOCH =>
+          if (onPartitionFenced(tp, latestEpochsForPartitions.get(tp).flatMap {
+            p =>
+              if (p.currentLeaderEpoch.isPresent) Some(p.currentLeaderEpoch.get())
+              else None
+          })) partitionsWithError += tp
+
+        case error =>
+          info(s"Retrying leaderEpoch request for partition $tp as the leader reported an error: $error")
+          partitionsWithError += tp
+      }
+    }
+
+    ResultWithPartitions(fetchOffsets, partitionsWithError)
+  }
+
+  /**
+   * remove the partition if the partition state is NOT updated. Otherwise, keep the partition active.
+   * @return true if the epoch in this thread is updated. otherwise, false
+   */
+  private def onPartitionFenced(tp: TopicPartition, requestEpoch: Option[Int]): Boolean = {
+    Option(partitionStates.stateValue(tp)).exists { currentFetchState =>
+      val currentLeaderEpoch = currentFetchState.currentLeaderEpoch
+      if (requestEpoch.contains(currentLeaderEpoch)) {
+        info(s"Partition $tp has an older epoch ($currentLeaderEpoch) than the current leader. Will await " +
+          s"the new LeaderAndIsr state before resuming fetching.")
+        markPartitionFailed(tp)
+        false
+      } else {
+        info(s"Partition $tp has an new epoch ($currentLeaderEpoch) than the current leader. retry the partition later")
+        true
+      }
+    }
+  }
+
+  /**
+   * processFetchRequest fetches data from the leader and returns whether the fetching logic should back off
+   * @param fetchStates
+   * @param fetchRequest
+   */
+  private def processFetchRequest(fetchStates: Map[TopicPartition, PartitionFetchState],
+                                  fetchRequest: FetchRequest.Builder): Boolean = {
+    val partitionsWithError = mutable.Set[TopicPartition]()
+    var responseData: Seq[(TopicPartition, FetchData)] = Seq.empty
+
+    try {
+      trace(s"Sending fetch request $fetchRequest")
+      responseData = fetchFromLeader(fetchRequest)
+    } catch {
+      case t: Throwable =>
+        // if (isRunning) { removing isRunning since it must be true
```

Comment:
Did you intend to leave the commented out condition and sleep here? I assume they aren't necessary anymore.

Reply:
You are right, the comment is not necessary, and I'll remove it.